### PR TITLE
Validate included types

### LINKED ISSUE #2192: Validation success when TSourceBase or TDestinationBase in IncludeBase<TSourceBase, TDestinationBase>() are not actual base types
### Source/destination classes
```csharp
class A { }
class A1 : A { }
class B { }

class ADto { }
class A1Dto : ADto { }
class BDto { }
```
### Mapping configuration
```csharp
CreateMap<A1, A1Dto>().IncludeBase<B, BDto>(); // B is not base class for A1 and BDto is not for A1Dto
```

### Expected behavior
Validation error if `TSourceBase` is not `TSource` or base class for `TSource` (same for `TDestinationBase` and `TDestination`), in this case, `B` is not base for `A1` and `BDto` is not base for `A1Dto`.

### Actual behavior
Validation successfully passed, target object successfully mapped.

### Steps to reproduce

```csharp
var config = new MapperConfiguration(cfg =>
{
    cfg.CreateMap<A1, A1Dto>().IncludeBase<B, BDto>();
});
config.AssertConfigurationIsValid();
var mapper = config.CreateMapper();

var source_a = new A1();
var target_a = mapper.Map<A1Dto>(source_a);
```

### PR DESCRIPTION
**Pull request by lbargaoanu**
Fixes #2192.

### CODE REVIEW

File: src/AutoMapper/Configuration/MappingExpression.cs [hunk 1]
```diff
@@ -239,22 +239,29 @@ private void IgnoreDestinationMember(MemberInfo property)
...
             where TOtherSource : TSource
             where TOtherDestination : TDestination 
-            => Include(typeof(TOtherSource), typeof(TOtherDestination));
+            => IncludeCore(typeof(TOtherSource), typeof(TOtherDestination));
 
         public IMappingExpression<TSource, TDestination> Include(Type otherSourceType, Type otherDestinationType)
+        {
+            CheckIsDerived(otherSourceType, typeof(TSource));
+            CheckIsDerived(otherDestinationType, typeof(TDestination));
+            return IncludeCore(otherSourceType, otherDestinationType); ;
```
Comment (jbogard) [inline src/AutoMapper/Configuration/MappingExpression.cs]: Whoops ;) ;)

File: src/AutoMapper/Configuration/MappingExpression.cs [hunk 2]
```diff
@@ -239,22 +239,29 @@ private void IgnoreDestinationMember(MemberInfo property)
...
+        IMappingExpression<TSource, TDestination> IncludeCore(Type otherSourceType, Type otherDestinationType)
         {
-            TypeMapActions.Add(tm => tm.IncludeDerivedTypes(otherSourceType, otherDestinationType));
-
+            TypeMapActions.Add(tm => tm.IncludeDerivedTypes(otherSourceType, otherDestinationType));
             return this;
-        }
+        }
 
         public IMappingExpression<TSource, TDestination> IncludeBase<TSourceBase, TDestinationBase>()
```
Comment (jbogard) [inline src/AutoMapper/Configuration/MappingExpression.cs]: Can we do the reverse here too? Where TSource : TSourceBase?

Comment (lbargaoanu) [inline src/AutoMapper/Configuration/MappingExpression.cs]: I've tried :) Unless I'm missing some syntactic stuff, we cannot.

File: src/AutoMapper/Configuration/MappingExpression.cs (hunk 1 above)
Comment (lbargaoanu) [inline src/AutoMapper/Configuration/MappingExpression.cs]: I don't see it :)

Comment (lbargaoanu) [inline src/AutoMapper/Configuration/MappingExpression.cs]: OK, _now_ I see it :)

File: src/AutoMapper/Configuration/MappingExpression.cs (hunk 2 above)
Comment (lbargaoanu) [inline src/AutoMapper/Configuration/MappingExpression.cs]: We need to constrain TSourceBase, not TSource, and there is no constraint is a base class of. That seems to be the idea.

Comment (jbogard) [inline src/AutoMapper/Configuration/MappingExpression.cs]: It seems we cannot bend the laws of C# to our will then.

Comment (lbargaoanu) [inline src/AutoMapper/Configuration/MappingExpression.cs]: Not today :)

**Pull request merged**